color/src/color.rs

```diff
@@ -0,0 +1,531 @@
+        HueDirection::Decreasing => *h2 -= 360.0 * dh.ceil(),
+    }
+}
+
+pub(crate) fn fixup_hues_for_interpolate(
+    a: [f32; 3],
+    b: &mut [f32; 3],
+    layout: ColorspaceLayout,
+    direction: HueDirection,
+) {
+    if let Some(ix) = layout.hue_channel() {
+        fixup_hue(a[ix], &mut b[ix], direction);
+    }
+}
+
+impl<T: Colorspace> OpaqueColor<T> {
+    pub const fn new(components: [f32; 3]) -> Self {
+        let cs = PhantomData;
+        Self { components, cs }
+    }
+
+    pub fn convert<U: Colorspace>(self) -> OpaqueColor<U> {
+        if TypeId::of::<T>() == TypeId::of::<U>() {
+            OpaqueColor::new(self.components)
+        } else {
+            let lin_rgb = T::to_linear_srgb(self.components);
+            OpaqueColor::new(U::from_linear_srgb(lin_rgb))
+        }
+    }
+
+    /// Add an alpha channel.
+    ///
+    /// This function is the inverse of [`AlphaColor::split`].
+    pub const fn with_alpha(self, alpha: f32) -> AlphaColor<T> {
+        AlphaColor::new(add_alpha(self.components, alpha))
+    }
+
+    /// Difference between two colors by Euclidean metric.
+    pub fn difference(self, other: Self) -> f32 {
+        let x = self.components;
+        let y = other.components;
+        let (d0, d1, d2) = (x[0] - y[0], x[1] - y[1], x[2] - y[2]);
+        (d0 * d0 + d1 * d1 + d2 * d2).sqrt()
+    }
+
+    /// Linearly interpolate colors, without hue fixup.
+    ///
+    /// This method produces meaningful results in rectangular colorspaces,
+    /// or if hue fixup has been applied.
+    #[must_use]
+    pub fn lerp_rect(self, other: Self, t: f32) -> Self {
+        self + t * (other - self)
+    }
+
+    /// Apply hue fixup for interpolation.
+    ///
+    /// Adjust the hue angle of `other` so that linear interpolation results in
+    /// the expected hue direction.
+    pub fn fixup_hues(self, other: &mut Self, direction: HueDirection) {
+        fixup_hues_for_interpolate(self.components, &mut other.components, T::LAYOUT, direction);
+    }
+
+    /// Linearly interpolate colors, with hue fixup if needed.
+    #[must_use]
+    pub fn lerp(self, mut other: Self, t: f32, direction: HueDirection) -> Self {
+        self.fixup_hues(&mut other, direction);
+        self.lerp_rect(other, t)
+    }
+
+    /// Scale the chroma by the given amount.
+    ///
+    /// See [`Colorspace::scale_chroma`] for more details.
+    #[must_use]
+    pub fn scale_chroma(self, scale: f32) -> Self {
+        Self::new(T::scale_chroma(self.components, scale))
+    }
+
+    /// Compute the relative luminance of the color.
+    ///
+    /// This can be useful for choosing contrasting colors, and follows the
+    /// WCAG 2.1 spec.
+    pub fn relative_luminance(self) -> f32 {
+        let rgb = T::to_linear_srgb(self.components);
+        0.2126 * rgb[0] + 0.7152 * rgb[1] + 0.0722 * rgb[2]
+    }
+}
+
+pub(crate) const fn split_alpha(x: [f32; 4]) -> ([f32; 3], f32) {
```

Comment:
This could potentially use `[x, y, z, a]: [f32; 4]`. This should only be a code-style thing.